### PR TITLE
Migrations, Testing, and a useful index

### PR DESCRIPTION
**Pull request by jpaddison3**
@jcmustin reviewed here: https://github.com/centre-for-effective-altruism/ForumAnalytics/pull/10

This PR attempts to clean up the analytics db a bit. First by adding migrations ([dbmate](https://github.com/amacneil/dbmate)) and testing ([pgtap](https://pgtap.org/)). Second by adding some useful indexes.

For LW to adopt this:
* First you'll probably want to figure out what branch I should actually be submitting this against.
* Then you'll need to run the migrations manually for now, by setting the postgres credentials in your .env and running dbmate. We should eventually have a github action that runs migrations on every merge to master, but we don't even have that for the main codebase yet, so I ain't concerned.

### CODE REVIEW

File: scripts/runDbTests.sh [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#! /bin/bash
+set -eox pipefail
+
+source .env
+
+PGPASSWORD=$PGPASSWORD bash -c "pg_prove -d forumanalytics -h $DATABASE_HOST -U postgres --verbose $1"
```
Comment (darkruby501) [inline scripts/runDbTests.sh]: So this would need to be changed to work for LessWrong.

File: db/migrations/20210824111234_author_metrics.sql [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    RETURN post_id;
+  END IF;
+  post_id := (SELECT (regexp_matches(path, '^/s/[a-zA-Z0-9]+/p/([a-zA-Z0-9]+)'))[1]);
+  RETURN post_id;
+END
+$$ LANGUAGE plpgsql IMMUTABLE STRICT;
+
+CREATE INDEX raw_event_type_post_id ON raw (event_type, get_post_id_from_path(event->>'path'));
+
+-- Timer event is the main way to tell how many people have read your post and
```
Comment (darkruby501) [inline db/migrations/20210824111234_author_metrics.sql:40]: It's kind of cool to see code that I very definitively wrote being used and built upon. I hope you tested the timer event carefully though. I never validated the data.

**Review by darkruby501 — APPROVED** [review]

File: db/migrations/20210824111234_author_metrics.sql (hunk 2 above)
Comment (jpaddison3) [reply db/migrations/20210824111234_author_metrics.sql:40]: Good to know.

File: scripts/runDbTests.sh (hunk 1 above)
Comment (jpaddison3) [reply scripts/runDbTests.sh]: Thinks for the catch

**Comment by jpaddison3** [comment]
Motion to merge.

(I'm pretty sure you don't have something set up to deploy automatically upon merge to master, but I think it might be a good idea to follow the policy of not clicking merge when the destination branch is master)

**Comment by darkruby501** [comment]
Merged. Sorry, I got confused about procedure. I should have just merged it before.

**Comment by jpaddison3** [comment]
I would actually advocate leaving it to me to say "motion to merge" if you have made comments I might want to address, like I did here.

**Comment by darkruby501** [comment]
That's fair. I didn't think the comment should be blocking, but happy to leave that decision to you.

**Comment by jpaddison3** [comment]
It wasn't *blocking* but it's easier to address if I can push to an existing PR

**Comment by darkruby501** [comment]
Duly noted.

**Pull request merged**